### PR TITLE
Add sequence numbers based optimistic concurrency control support to Engine

### PR DESCRIPTION
**Pull request by bleskes**
This PR add support to engine operations for resolving and verifying the sequence number and primary term of the last modification to a document before performing an operation. This is infrastructure to move our [optimistic concurrency control](http://en.wikipedia.org/wiki/Optimistic_concurrency_control) API to use sequence numbers instead of internal versioning.

Relates #36148 
Relates #10708

PS there are still discussions about the parameter naming. I'll rename these once we have settled on something.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDVersionAndSeqNoLookup.java [hunk 1]
```diff
@@ -108,7 +108,27 @@ public DocIdAndVersion lookupVersion(BytesRef id, LeafReaderContext context)
             if (versions.advanceExact(docID) == false) {
                 throw new IllegalArgumentException("Document [" + docID + "] misses the [" + VersionFieldMapper.NAME + "] field");
             }
-            return new DocIdAndVersion(docID, versions.longValue(), context.reader(), context.docBase);
+            final long seqNo;
+            final long term;
+            if (loadSeqNo) {
+                NumericDocValues seqNos = context.reader().getNumericDocValues(SeqNoFieldMapper.NAME);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDVersionAndSeqNoLookup.java]: we should clean this up in master once backported. can you put a comment in there?

**Review by s1monw — APPROVED** [review]
LGTM - I am not happy with the naming. I think `cas` should not be in there. Can we just name them `expectedSeqId` `expectedPrimaryTerm` for now? That's what for instance `Integer.compareAndSet` uses as it's parameter names.

**Review by dnhatn — APPROVED** [review]
LGTM.

**Comment by bleskes** [comment]
Thanks @s1monw and @dnhatn. I incorporated your suggestion.

**Comment by bleskes** [comment]
run gradle build tests 1

**Comment by bleskes** [comment]
The build failure was regarding refresh listeners. I don't think it's a related but probably signals a concurrency bug. I tried to reproduce this without success (I run ~300 time). Capturing the info here in case it's needed.

```
15:42:40 FAILURE 18.0s J0 | RefreshListenersTests.testConcurrentRefresh <<< FAILURES!
15:42:40    > Throwable #1: java.lang.AssertionError
15:42:40    > 	at __randomizedtesting.SeedInfo.seed([2848E94A583B5A6E:EF0293D9D997F804]:0)
15:42:40    > 	at org.elasticsearch.index.shard.RefreshListenersTests.lambda$testConcurrentRefresh$8(RefreshListenersTests.java:282)
15:42:40    > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:847)
15:42:40    > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:821)
15:42:40    > 	at org.elasticsearch.index.shard.RefreshListenersTests.testConcurrentRefresh(RefreshListenersTests.java:282)
```

```
./gradlew :server:test -Dtests.seed=2848E94A583B5A6E -Dtests.class=org.elasticsearch.index.shard.RefreshListenersTests -Dtests.method="testConcurrentRefresh" -Dtests.security.manager=true -Dtests.locale=pt-PT -Dtests.timezone=UCT -Dcompiler.java=11 -Druntime.java=8
```

**Comment by ywelsch** [comment]
> The build failure was regarding refresh listeners. I don't think it's a related

Similar assertion recently failed on master (https://github.com/elastic/elasticsearch/issues/24418#issuecomment-446601536), so looks to be unrelated to your changes here

**Comment by bleskes** [comment]
run gradle build tests 1

**Pull request merged**